File: archive/cdk-ts/Pr5857/lib/pipeline-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as codepipeline from 'aws-cdk-lib/aws-codepipeline';
import * as codepipeline_actions from 'aws-cdk-lib/aws-codepipeline-actions';
import * as codecommit from 'aws-cdk-lib/aws-codecommit';
import * as codebuild from 'aws-cdk-lib/aws-codebuild';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as sns from 'aws-cdk-lib/aws-sns';
import { Construct } from 'constructs';

export interface PipelineStackProps extends cdk.StackProps {
  environmentSuffix: string;
  environmentName: string;
  repositoryName: string;
  notificationTopic: sns.ITopic;
}

export class PipelineStack extends cdk.Stack {
  public readonly pipeline: codepipeline.Pipeline;

  constructor(scope: Construct, id: string, props: PipelineStackProps) {
    super(scope, id, props);

    const {
      environmentSuffix,
      environmentName,
      repositoryName,
      notificationTopic,
    } = props;

    // CodeCommit repository
    const repository = codecommit.Repository.fromRepositoryName(
      this,
      'Repository',
      repositoryName
    );

    // S3 bucket for pipeline artifacts
    const artifactBucket = new s3.Bucket(
      this,
      `PipelineArtifacts-${environmentSuffix}`,
      {
        bucketName: `pipeline-artifacts-${environmentSuffix}-${cdk.Aws.ACCOUNT_ID}`,
        encryption: s3.BucketEncryption.S3_MANAGED,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
        autoDeleteObjects: true,
      }
    );

    // CodeBuild project for CDK synth
    const buildProject = new codebuild.PipelineProject(
      this,
      `BuildProject-${environmentSuffix}`,
      {
        projectName: `migration-build-${environmentSuffix}`,
        environment: {
          buildImage: codebuild.LinuxBuildImage.STANDARD_7_0,
          privileged: true,
        },
        buildSpec: codebuild.BuildSpec.fromObject({
          version: '0.2',
          phases: {
            install: {
              commands: ['npm install -g aws-cdk', 'npm ci'],
            },
            build: {
              commands: ['npm run build', 'npm run test', 'cdk synth'],
            },
          },
          artifacts: {
            'base-directory': 'cdk.out',
            files: '**/*',
          },
        }),
      }
    );

    // Pipeline
    const sourceOutput = new codepipeline.Artifact('SourceOutput');
    const buildOutput = new codepipeline.Artifact('BuildOutput');

    this.pipeline = new codepipeline.Pipeline(
      this,
      `Pipeline-${environmentSuffix}`,
      {
        pipelineName: `migration-pipeline-${environmentSuffix}`,
        artifactBucket: artifactBucket,
        restartExecutionOnUpdate: true,
      }
    );

    // Source stage
    this.pipeline.addStage({
      stageName: 'Source',
      actions: [
        new codepipeline_actions.CodeCommitSourceAction({
          actionName: 'CodeCommit',
          repository: repository,
          branch: environmentName === 'prod' ? 'main' : environmentName,
          output: sourceOutput,
        }),
      ],
    });

    // Build stage
    this.pipeline.addStage({
      stageName: 'Build',
      actions: [
        new codepipeline_actions.CodeBuildAction({
          actionName: 'CDK_Build',
          project: buildProject,
          input: sourceOutput,
          outputs: [buildOutput],
        }),
      ],
    });

    // Manual approval for production
    if (environmentName === 'prod') {
      this.pipeline.addStage({
        stageName: 'Approval',
        actions: [
          new codepipeline_actions.ManualApprovalAction({
            actionName: 'ManualApproval',
            additionalInformation:
              'Please review changes before deploying to production',
            notificationTopic: notificationTopic,
          }),
        ],
      });
    }

    // Deploy stage
    this.pipeline.addStage({
      stageName: 'Deploy',
      actions: [
        new codepipeline_actions.CloudFormationCreateUpdateStackAction({
          actionName: 'Deploy_Infrastructure',
          stackName: `MigrationStack-${environmentSuffix}`,
          templatePath: buildOutput.atPath('MigrationStack.template.json'),
          adminPermissions: true,
          parameterOverrides: {
            EnvironmentSuffix: environmentSuffix,
          },
        }),
      ],
    });

    new cdk.CfnOutput(this, 'PipelineName', {
      value: this.pipeline.pipelineName,
      description: 'CodePipeline Name',
      exportName: `PipelineName-${environmentSuffix}`,
    });

    new cdk.CfnOutput(this, 'PipelineArn', {
      value: this.pipeline.pipelineArn,
      description: 'CodePipeline ARN',
    });
  }
}

File: archive/cdk-go/Pr5955/lib/pipeline-stack.go
package lib

import (
	"fmt"

	"github.com/aws/aws-cdk-go/awscdk/v2"
	"github.com/aws/aws-cdk-go/awscdk/v2/awscodebuild"
	"github.com/aws/aws-cdk-go/awscdk/v2/awscodepipeline"
	"github.com/aws/aws-cdk-go/awscdk/v2/awscodepipelineactions"
	"github.com/aws/aws-cdk-go/awscdk/v2/awss3"
	"github.com/aws/constructs-go/constructs/v10"
	"github.com/aws/jsii-runtime-go"
)

type PipelineStackProps struct {
	awscdk.StackProps
	EnvironmentSuffix string
	Environment       string
	RepositoryName    string
	BranchName        string
}

func NewPipelineStack(scope constructs.Construct, id string, props *PipelineStackProps) awscdk.Stack {
	var sprops awscdk.StackProps
	if props != nil {
		sprops = props.StackProps
	}
	stack := awscdk.NewStack(scope, &id, &sprops)

	// Create artifact bucket for pipeline
	artifactBucket := awss3.NewBucket(stack, jsii.String(fmt.Sprintf("pipeline-artifacts-%s", props.EnvironmentSuffix)), &awss3.BucketProps{
		BucketName:        jsii.String(fmt.Sprintf("pipeline-artifacts-%s", props.EnvironmentSuffix)),
		Versioned:         jsii.Bool(true),
		RemovalPolicy:     awscdk.RemovalPolicy_DESTROY,
		AutoDeleteObjects: jsii.Bool(true),
		Encryption:        awss3.BucketEncryption_S3_MANAGED,
	})

	// Create source output artifact
	sourceOutput := awscodepipeline.NewArtifact(jsii.String("SourceOutput"), nil)

	// Create build project for CDK synth
	buildProject := awscodebuild.NewPipelineProject(stack, jsii.String(fmt.Sprintf("build-project-%s", props.EnvironmentSuffix)), &awscodebuild.PipelineProjectProps{
		ProjectName: jsii.String(fmt.Sprintf("payment-cdk-build-%s", props.EnvironmentSuffix)),
		Environment: &awscodebuild.BuildEnvironment{
			BuildImage:  awscodebuild.LinuxBuildImage_STANDARD_7_0(),
			ComputeType: awscodebuild.ComputeType_SMALL,
		},
		BuildSpec: awscodebuild.BuildSpec_FromObject(&map[string]interface{}{
			"version": "0.2",
			"phases": map[string]interface{}{
				"install": map[string]interface{}{
					"commands": []string{
						"npm install -g aws-cdk",
						"go version",
					},
				},
				"build": map[string]interface{}{
					"commands": []string{
						"go mod download",
						"cdk synth",
					},
				},
			},
			"artifacts": map[string]interface{}{
				"base-directory": "cdk.out",
				"files": []string{
					"**/*",
				},
			},
		}),
	})

	// Create build output artifact
	buildOutput := awscodepipeline.NewArtifact(jsii.String("BuildOutput"), nil)

	// Create build action
	buildAction := awscodepipelineactions.NewCodeBuildAction(&awscodepipelineactions.CodeBuildActionProps{
		ActionName: jsii.String("Build"),
		Project:    buildProject,
		Input:      sourceOutput,
		Outputs:    &[]awscodepipeline.Artifact{buildOutput},
	})

	// Create manual approval action for production
	var approvalAction awscodepipelineactions.ManualApprovalAction
	if props.Environment == "prod" {
		approvalAction = awscodepipelineactions.NewManualApprovalAction(&awscodepipelineactions.ManualApprovalActionProps{
			ActionName:            jsii.String("ManualApproval"),
			AdditionalInformation: jsii.String("Please review the changes before deploying to production"),
			RunOrder:              jsii.Number(1),
		})
	}

	// Create deploy action
	deployAction := awscodepipelineactions.NewCloudFormationCreateUpdateStackAction(&awscodepipelineactions.CloudFormationCreateUpdateStackActionProps{
		ActionName:       jsii.String("Deploy"),
		StackName:        jsii.String(fmt.Sprintf("PaymentStack-%s", props.Environment)),
		TemplatePath:     buildOutput.AtPath(jsii.String(fmt.Sprintf("PaymentStack-%s.template.json", props.Environment))),
		AdminPermissions: jsii.Bool(true),
		RunOrder:         jsii.Number(2),
	})

	// Create pipeline stages
	var stages []*awscodepipeline.StageProps

	// Source stage (placeholder - in real implementation would use CodeCommit or GitHub)
	stages = append(stages, &awscodepipeline.StageProps{
		StageName: jsii.String("Source"),
		Actions:   &[]awscodepipeline.IAction{},
	})

	// Build stage
	stages = append(stages, &awscodepipeline.StageProps{
		StageName: jsii.String("Build"),
		Actions:   &[]awscodepipeline.IAction{buildAction},
	})

	// Deploy stage with optional approval
	var deployStageActions []awscodepipeline.IAction
	if props.Environment == "prod" {
		deployStageActions = []awscodepipeline.IAction{approvalAction, deployAction}
	} else {
		deployStageActions = []awscodepipeline.IAction{deployAction}
	}

	stages = append(stages, &awscodepipeline.StageProps{
		StageName: jsii.String("Deploy"),
		Actions:   &deployStageActions,
	})

	// Create pipeline
	pipeline := awscodepipeline.NewPipeline(stack, jsii.String(fmt.Sprintf("pipeline-%s", props.EnvironmentSuffix)), &awscodepipeline.PipelineProps{
		PipelineName:             jsii.String(fmt.Sprintf("payment-pipeline-%s", props.EnvironmentSuffix)),
		ArtifactBucket:           artifactBucket,
		Stages:                   &stages,
		RestartExecutionOnUpdate: jsii.Bool(false),
	})

	// Tag pipeline
	awscdk.Tags_Of(pipeline).Add(jsii.String("Environment"), jsii.String(props.Environment), nil)
	awscdk.Tags_Of(pipeline).Add(jsii.String("CostCenter"), jsii.String("PaymentProcessing"), nil)

	// Output pipeline ARN
	awscdk.NewCfnOutput(stack, jsii.String("PipelineArn"), &awscdk.CfnOutputProps{
		Value:       pipeline.PipelineArn(),
		Description: jsii.String("CodePipeline ARN"),
		ExportName:  jsii.String(fmt.Sprintf("pipeline-arn-%s", props.EnvironmentSuffix)),
	})

	return stack
}
